lib/kafka/sasl_plain_authenticator.rb

```diff
@@ -0,0 +1,37 @@
+module Kafka
+  class SaslPlainAuthenticator
+    PLAIN_IDENT = "PLAIN"
+
+    def initialize(connection:, logger:, authzid:, username:, password:)
+      @connection = connection
+      @logger = logger
+      @authzid = authzid
+      @username = username
+      @password = password
+    end
+
+    def authenticate!
+      response = @connection.send_request(Kafka::Protocol::SaslHandshakeRequest.new(PLAIN_IDENT))
+
+      @encoder = @connection.encoder
+      @decoder = @connection.decoder
+
+      unless response.error_code == 0 && response.enabled_mechanisms.include?(PLAIN_IDENT)
+        raise Kafka::Error, "#{PLAIN_IDENT} is not supported."
+      end
+
+      # SASL PLAIN
+      msg = [@authzid.to_s,
+             @username.to_s,
+             @password.to_s].join("\000").force_encoding("utf-8")
+      @encoder.write([msg.bytesize].pack("l>")+msg)
```

Comment:
`#write_bytes` should do exactly what you're doing here (https://github.com/zendesk/ruby-kafka/blob/master/lib/kafka/protocol/encoder.rb#L90), with the only difference being when `msg` is `nil` – which it definitely shouldn't be here. Can you try again?